tests/components/waterkotte_heatpump/test_config_flow.py

```diff
@@ -0,0 +1,158 @@
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "pywaterkotte.ecotouch.Ecotouch.login",
+        side_effect=AuthenticationException,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                "host": "1.1.1.1",
+                "username": "test-username",
+                "password": "test-password",
+            },
+        )
+
+    assert result2["type"] == FlowResultType.FORM
+    assert result2["errors"] == {"base": "invalid_auth"}
+
+
+async def test_form_cannot_connect(hass: HomeAssistant) -> None:
+    """Test we handle cannot connect error."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "pywaterkotte.ecotouch.Ecotouch.login",
+        side_effect=ConnectionException,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                "host": "1.1.1.1",
+                "username": "test-username",
+                "password": "test-password",
+            },
+        )
+
+    assert result2["type"] == FlowResultType.FORM
+    assert result2["errors"] == {"base": "cannot_connect"}
+
+
+async def test_device_already_configured(hass: HomeAssistant) -> None:
+    """Test if we find out if device is configured twice."""
+    entry = MockConfigEntry(
+        domain=DOMAIN, unique_id="serial_no_123", data={"host": "1.1.1.1"}
+    )
+    entry.add_to_hass(hass)
+
+    with patch(
+        "pywaterkotte.ecotouch.Ecotouch.login",
+        return_value=True,
+    ), patch(
+        "pywaterkotte.ecotouch.Ecotouch.read_value",
+        side_effect=lambda tag: {
+            EcotouchTags.HEATPUMP_TYPE: 42,
+            EcotouchTags.SERIAL_NUMBER: "serial_no_123",
+        }[tag],
+    ), patch(
+        "pywaterkotte.ecotouch.Ecotouch.decode_heatpump_series",
+        return_value="heatpump type",
+    ):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": config_entries.SOURCE_USER}
+        )
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                "host": "1.2.3.4",
+                "username": "test-user1",
+                "password": "test-pass2",
+            },
+        )
+
+    await hass.async_block_till_done()
+    assert result2["type"] == FlowResultType.ABORT
+    assert result2["reason"] == "already_configured"
+
+
+async def test_unexpected_exception(hass: HomeAssistant) -> None:
+    """Test we handle unexpected exceptions."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "pywaterkotte.ecotouch.Ecotouch.login",
```

Comment:
instead of patching it globally, you are supposed to patch it from within the component.
"homeassistant.components.waterkotte_heatpump. ..."